src/main/java/org/jvnet/hudson/test/InboundAgentRule.java

```diff
@@ -50,32 +52,184 @@
+             * Build and return an {@link Options}.
+             *
+             * @return a new {@link Options}
+             */
+            Options build();
+        }
+
+        private static class BuilderImpl implements Builder {
+
+            private final Options options = new Options();
+
+            @Override
+            public Builder name(String name) {
+                options.setName(name);
+                return this;
+            }
+
+            @Override
+            public Builder secret() {
+                options.setSecret(true);
+                return this;
+            }
+
+            @Override
+            public Builder webSocket() {
+                options.setWebSocket(true);
+                return this;
+            }
+
+            @Override
+            public Builder skipStart() {
+                options.setStart(false);
+                return this;
+            }
+
+            @Override
+            public Options build() {
+                return options;
+            }
+        }
+
+        public static Builder newBuilder() {
+            return new BuilderImpl();
+        }
+    }
 
     /**
      * Creates, attaches, and starts a new inbound agent.
+     *
      * @param name an optional {@link Slave#getNodeName}
      */
-    @SuppressFBWarnings(value = "PATH_TRAVERSAL_IN", justification = "just for test code")
     public Slave createAgent(@NonNull JenkinsRule r, @CheckForNull String name) throws Exception {
-        if (name == null) {
-            name = "agent" + r.jenkins.getNodes().size();
+        return createAgent(r, Options.newBuilder().name(name).build());
+    }
+
+    /**
+     * Creates, attaches, and optionally starts a new inbound agent.
+     *
+     * @param options the options
+     */
+    @SuppressFBWarnings(value = {"PATH_TRAVERSAL_IN", "NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE"}, justification = "just for test code")
+    public Slave createAgent(@NonNull JenkinsRule r, Options options) throws Exception {
+        if (options.getName() == null) {
+            options.setName("agent" + r.jenkins.getNodes().size());
         }
-        // TODO 2.216+ support WebSocket option
-        DumbSlave s = new DumbSlave(name, new File(r.jenkins.getRootDir(), "agent-work-dirs/" + name).getAbsolutePath(), new JNLPLauncher(true));
-        s.setNumExecutors(1); // TODO 2.234+ delete (default switched from 2 to 1)
+        JNLPLauncher launcher = new JNLPLauncher(true);
+        launcher.setWebSocket(options.isWebSocket());
+        DumbSlave s = new DumbSlave(options.getName(), new File(r.jenkins.getRootDir(), "agent-work-dirs/" + options.getName()).getAbsolutePath(), launcher);
         s.setRetentionStrategy(RetentionStrategy.NOOP);
         r.jenkins.addNode(s);
-        start(r, name);
+        // SlaveComputer#_connect runs asynchronously. Wait for it to finish for a more deterministic test.
+        while (s.toComputer().getOfflineCause() == null) {
+            Thread.sleep(100);
```

Comment:
is there a chance this never happens and we end up with an infinite loop? I guess the test timeout would kick in eventually.